File: SDE_sheet/Linked-List/07_intersection/solution.scala
/**
 * Definition for singly-linked list.
 * class ListNode(var _x: Int = 0) {
 *   var next: ListNode = null
 *   var x: Int = _x
 * }
 */

object Solution {
    def getIntersectionNode(headA: ListNode, headB: ListNode): ListNode = {
        var a = headA
        var b = headB
        while (a != b) {
        a = if (a != null) a.next else headB
        b = if (b != null) b.next else headA
        }
        a
    }
}

File: SDE_sheet/Linked-List/07_intersection/solution.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func getIntersectionNode(_ headA: ListNode?, _ headB: ListNode?) -> ListNode? {
        var a = headA
        var b = headB
        while a !== b {
            a = a != nil ? a?.next : headB
            b = b != nil ? b?.next : headA
        }
        return a
    }
}
